File: app/schemas/super_admin.py
from datetime import date, datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, Field

from app.schemas.common import ORMModel


StatusPlataforma = Literal["TRIAL", "ATIVA", "SUSPENSA", "CANCELADA", "ARQUIVADA"]


class SuperAdminLoginRequest(BaseModel):
    email: str = Field(min_length=3, max_length=150)
    senha: str = Field(min_length=8, max_length=128)


class SuperAdminTokenResponse(BaseModel):
    access_token: str
    token_type: str = "bearer"
    expires_in: int


class SuperAdminOut(ORMModel):
    id: int
    nome: str
    email: str
    ativo: bool
    dois_fatores_ativo: bool
    ultimo_login: datetime | None
    created_at: datetime


class SuperAdminAlterarSenhaRequest(BaseModel):
    senha_atual: str = Field(min_length=8, max_length=128)
    nova_senha: str = Field(min_length=8, max_length=128)


class PlanoBase(BaseModel):
    nome: str = Field(min_length=2, max_length=60)
    descricao: str | None = None
    preco: float = Field(ge=0)
    preco_anual: float | None = Field(default=None, ge=0)
    ativo: bool = True
    periodo_teste_dias: int = Field(default=14, ge=0, le=90)
    limite_usuarios: int | None = Field(default=None, ge=1)
    limite_clientes: int | None = Field(default=None, ge=1)
    limite_agendamentos_mes: int | None = Field(default=None, ge=1)
    limite_conversas_mes: int | None = Field(default=None, ge=1)
    limite_mensagens_ia_mes: int | None = Field(default=None, ge=0)
    limite_canais: int | None = Field(default=None, ge=1)
    limite_armazenamento_mb: int | None = Field(default=None, ge=1)
    ia_incluida: bool = False
    ia_adicional_disponivel: bool = True
    recursos: dict[str, bool] = Field(default_factory=dict)


class PlanoCreate(PlanoBase):
    codigo: str = Field(min_length=2, max_length=40, pattern=r"^[A-Z0-9_]+$")


class PlanoUpdate(BaseModel):
    nome: str | None = Field(default=None, min_length=2, max_length=60)
    descricao: str | None = None
    preco: float | None = Field(default=None, ge=0)
    preco_anual: float | None = Field(default=None, ge=0)
    ativo: bool | None = None
    periodo_teste_dias: int | None = Field(default=None, ge=0, le=90)
    limite_usuarios: int | None = Field(default=None, ge=1)
    limite_clientes: int | None = Field(default=None, ge=1)
    limite_agendamentos_mes: int | None = Field(default=None, ge=1)
    limite_conversas_mes: int | None = Field(default=None, ge=1)
    limite_mensagens_ia_mes: int | None = Field(default=None, ge=0)
    limite_canais: int | None = Field(default=None, ge=1)
    limite_armazenamento_mb: int | None = Field(default=None, ge=1)
    ia_incluida: bool | None = None
    ia_adicional_disponivel: bool | None = None
    recursos: dict[str, bool] | None = None


class PlanoOut(PlanoBase):
    id: int
    codigo: str
    created_at: datetime
    updated_at: datetime


class EmpresaSuperAdminCreate(BaseModel):
    nome: str = Field(min_length=2, max_length=150)
    cnpj: str = Field(min_length=11, max_length=18)
    telefone: str | None = Field(default=None, max_length=20)
    email: str | None = Field(default=None, max_length=150)
    cidade: str | None = Field(default=None, max_length=100)
    estado: str | None = Field(default=None, min_length=2, max_length=2)
    timezone: str = Field(default="America/Sao_Paulo", max_length=50)
    plano_id: int = Field(gt=0)
    periodo_teste_dias: int = Field(default=14, ge=0, le=90)
    admin_nome: str = Field(min_length=2, max_length=150)
    admin_email: str = Field(min_length=3, max_length=150)
    admin_senha: str = Field(min_length=8, max_length=128)


class EmpresaSuperAdminUpdate(BaseModel):
    plano_id: int | None = Field(default=None, gt=0)
    status: StatusPlataforma | None = None
    trial_fim: date | None = None
    recursos_personalizados: dict[str, bool] | None = None
    limites_personalizados: dict[str, int | None] | None = None
    ia_adicional_ativo: bool | None = None
    ia_limite_adicional: int | None = Field(default=None, ge=0)
    observacoes: str | None = None


class EmpresaResumoOut(BaseModel):
    id: int
    nome: str
    cnpj: str
    email: str | None
    cidade: str | None
    estado: str | None
    ativo: bool
    status: StatusPlataforma
    plano_id: int | None
    plano_nome: str | None
    trial_fim: date | None
    usuarios_ativos: int
    clientes: int
    agendamentos_mes: int
    conversas_mes: int
    ia_adicional_ativo: bool
    created_at: datetime


class UsoEmpresaOut(BaseModel):
    usuarios_ativos: int
    clientes: int
    agendamentos_mes: int
    conversas_mes: int
    canais_ativos: int
    mensagens_ia_mes: int
    limites: dict[str, int | None]
    recursos: dict[str, bool]


class EmpresaDetalheOut(BaseModel):
    id: int
    nome: str
    cnpj: str
    telefone: str | None
    email: str | None
    cidade: str | None
    estado: str | None
    timezone: str
    ativo: bool
    status: StatusPlataforma
    plano_id: int | None
    plano_nome: str | None
    trial_fim: date | None
    recursos_personalizados: dict[str, bool]
    limites_personalizados: dict[str, int | None]
    ia_adicional_ativo: bool
    ia_limite_adicional: int
    observacoes: str | None
    uso: UsoEmpresaOut
    created_at: datetime
    updated_at: datetime


class ConfigIASuperAdminUpdate(BaseModel):
    nome_assistente: str = Field(min_length=2, max_length=80)
    mensagem_boas_vindas: str | None = None
    prompt: str | None = None
    temperatura: Decimal = Field(ge=0, le=2, max_digits=3, decimal_places=2)


class ConfigIASuperAdminOut(ORMModel):
    id: int
    empresa_id: int
    nome_assistente: str
    mensagem_boas_vindas: str | None
    prompt: str | None
    temperatura: Decimal


class DashboardSuperAdminOut(BaseModel):
    empresas_total: int
    empresas_ativas: int
    empresas_trial: int
    empresas_suspensas: int
    usuarios_ativos: int
    agendamentos_mes: int
    conversas_mes: int
    planos_ativos: int
    empresas_por_plano: list[dict[str, int | str]]
    alertas: list[dict[str, str]]


class SuperAdminLogOut(ORMModel):
    id: int
    super_admin_id: int | None
    empresa_id: int | None
    acao: str
    entidade: str | None
    entidade_id: int | None
    dados_anteriores: dict | None
    dados_novos: dict | None
    ip: str | None
    created_at: datetime

File: frontend/src/types/superAdmin.ts
export type StatusPlataforma =
  | "TRIAL"
  | "ATIVA"
  | "SUSPENSA"
  | "CANCELADA"
  | "ARQUIVADA";

export interface SuperAdminUsuario {
  id: number;
  nome: string;
  email: string;
  ativo: boolean;
  dois_fatores_ativo: boolean;
  ultimo_login: string | null;
  created_at: string;
}

export interface SuperAdminDashboard {
  empresas_total: number;
  empresas_ativas: number;
  empresas_trial: number;
  empresas_suspensas: number;
  usuarios_ativos: number;
  agendamentos_mes: number;
  conversas_mes: number;
  planos_ativos: number;
  empresas_por_plano: Array<{ plano: string; empresas: number }>;
  alertas: Array<{ tipo: string; titulo: string; mensagem: string }>;
}

export interface SuperAdminFinancialDashboard {
  start_date: string;
  end_date: string;
  companies_total: number;
  companies_active: number;
  companies_trial: number;
  companies_suspended: number;
  companies_overdue: number;
  new_companies_period: number;
  active_users: number;
  appointments_period: number;
  conversations_period: number;
  active_plans: number;
  active_ai_addons: number;
  estimated_mrr: string | number;
  estimated_arr: string | number;
  new_contracts_period: number;
  new_contracts_monthly_value: string | number;
  audit_events_period: number;
  companies_by_plan: Array<{ plano: string; empresas: number }>;
  alerts: Array<{ type: string; title: string; message: string }>;
  recent_audit: Array<{
    id: number;
    action: string;
    entity: string | null;
    company_id: number | null;
    created_at: string;
  }>;
}

export interface PlanoSuperAdmin {
  id: number;
  codigo: string;
  nome: string;
  descricao: string | null;
  preco: string | number;
  preco_anual: string | number | null;
  ativo: boolean;
  periodo_teste_dias: number;
  limite_usuarios: number | null;
  limite_clientes: number | null;
  limite_agendamentos_mes: number | null;
  limite_conversas_mes: number | null;
  limite_mensagens_ia_mes: number | null;
  limite_canais: number | null;
  limite_armazenamento_mb: number | null;
  ia_incluida: boolean;
  ia_adicional_disponivel: boolean;
  recursos: Record<string, boolean>;
  created_at: string;
  updated_at: string;
}

export interface EmpresaSuperAdminResumo {
  id: number;
  nome: string;
  cnpj: string;
  email: string | null;
  cidade: string | null;
  estado: string | null;
  ativo: boolean;
  status: StatusPlataforma;
  plano_id: number | null;
  plano_nome: string | null;
  trial_fim: string | null;
  usuarios_ativos: number;
  clientes: number;
  agendamentos_mes: number;
  conversas_mes: number;
  ia_adicional_ativo: boolean;
  created_at: string;
}

export interface UsoEmpresaSuperAdmin {
  usuarios_ativos: number;
  clientes: number;
  agendamentos_mes: number;
  conversas_mes: number;
  canais_ativos: number;
  mensagens_ia_mes: number;
  limites: Record<string, number | null>;
  recursos: Record<string, boolean>;
}

export interface EmpresaSuperAdminDetalhe {
  id: number;
  nome: string;
  cnpj: string;
  telefone: string | null;
  email: string | null;
  cidade: string | null;
  estado: string | null;
  timezone: string;
  ativo: boolean;
  status: StatusPlataforma;
  plano_id: number | null;
  plano_nome: string | null;
  trial_fim: string | null;
  recursos_personalizados: Record<string, boolean>;
  limites_personalizados: Record<string, number | null>;
  ia_adicional_ativo: boolean;
  ia_limite_adicional: number;
  observacoes: string | null;
  uso: UsoEmpresaSuperAdmin;
  created_at: string;
  updated_at: string;
}

export interface ConfigIASuperAdmin {
  id: number;
  empresa_id: number;
  nome_assistente: string;
  mensagem_boas_vindas: string | null;
  prompt: string | null;
  temperatura: string | number;
}

export interface SuperAdminLog {
  id: number;
  super_admin_id: number | null;
  empresa_id: number | null;
  acao: string;
  entidade: string | null;
  entidade_id: number | null;
  dados_anteriores: Record<string, unknown> | null;
  dados_novos: Record<string, unknown> | null;
  ip: string | null;
  created_at: string;
}

export interface SuperAdminOutletContext {
  usuario: SuperAdminUsuario;
  atualizarUsuario: () => Promise<void>;
}
